syntax = "proto3";

option java_package = "ca.eandb.jmist.proto";
option java_outer_classname = "MeshProtos";

message VectorSlice {
  enum Format {
    DOUBLE_XYZ = 0;
  }
  Format format = 1;
  int64 offset = 2;
}

message IndexSlice {
  enum Format {
    UINT8 = 0;
    UINT16 = 1;
    UINT32 = 2;
  }
  Format format = 1;
  int64 offset = 2;
}

message UVSlice {
  enum Format {
    DOUBLE_UV = 0;
  }
  Format format = 1;
  int32 offset = 2;
}

message ColorSlice {
  enum Format {
    DOUBLE_RGB = 0;
    DOUBLE_RGBA = 1;
    DOUBLE_XYZ = 2;
  }
  Format format = 1;
  int64 offset = 2;
}

message VertexBlock {
  int64 offset = 1;
  int64 stride = 2;
  int32 count = 3;
  VectorSlice coords = 4;
  VectorSlice normals = 5;
  UVSlice uvs = 6;
  ColorSlice colors = 7;
}

message LoopBlock {
  int64 offset = 1;
  int64 stride = 2;
  int32 count = 3;
  IndexSlice indices = 4;
  VectorSlice normals = 5;
  UVSlice uvs = 6;
  ColorSlice colors = 7;
}

message FaceBlock {
  int64 offset = 1;
  int64 stride = 2;
  int32 count = 3;
  IndexSlice loop_start = 4;
  IndexSlice loop_count = 5;
  IndexSlice materials = 6;
  ColorSlice colors = 7;
}

message MeshFormat {
  uint64 offset = 1;
  uint32 max_polygon_vertex_count = 2;

  VertexBlock vertices = 3;
  LoopBlock loops = 4;
  FaceBlock faces = 5;
}

message Mesh {
  MeshFormat format = 1;
  bytes data = 2;
}
